### DISCUSSION TITLE
My doubts in some test cases

### DISCUSSION THREAD

**Discussion by Aaron-cly** [CastleLab/COMP3021-2021Fall-PA1-Student-Version#4] [Q&A] [answered]
In some test cases, gameboard is instantiated and the constructor throws exception because not all gems are reachable. Therefore, the tests are not properly executed.

* These are the tests that I have doubts in
<img width="769" alt="Screen Shot 2021-09-07 at 18 30 58 PM" src="https://user-images.githubusercontent.com/68986494/132330269-470035c6-57a9-4a49-aea5-23068ebdaa4c.png">

* If I do not check whether all gems are reachable in my code, you can see that the tests above all passed except the test checking for unreachable gems
<img width="699" alt="Screen Shot 2021-09-07 at 18 32 19 PM" src="https://user-images.githubusercontent.com/68986494/132330446-491c343f-995a-4eb9-82d5-30ea76126c55.png">

* Is this result expected? I see that the provided tests are a part of the grading scheme. Thank you.

**Comment by Derppening** [comment]
With my reference implementation, this does not occur.

For instance, with `GameBoardTest.Get Cell - int-overload`, if you look at the source code:

```java
        final var cells = GameBoardUtils.createEmptyCellArray(rows, cols, EntityCell::new);

        ((EntityCell) cells[0][0]).setEntity(new Player());
        ((EntityCell) cells[1][1]).setEntity(new Gem());

        gameBoard = new GameBoard(rows, cols, cells);
```

This creates a gameboard which looks like this:

```
P..
.*.
...
```

You can see that the gem is reachable, so it shouldn't throw an exception. Therefore, it is more likely that your implementation for "all-gems reachable" is incorrect.

For your reference, with my reference implementation, only `GameBoardTest.Instance Creation - Unreachable Gem in Linear Path` will fail if the "all-gems unreachable" implementation is missing.

*Edit:* Remember that the provided test cases are in no way comprehensive - You need to do your own testing to make sure that implementing something doesn't break everything else.

**Reply by Aaron-cly** [reply]
Thank you for the reply. I also checked the test implementation and I wondered if the player can really get to the middle gem coz it can at best circle around it. Am I getting anything wrong here?

**Reply by Derppening** [reply]
I can move, say: `RIGHT`/`DOWN` or `DOWN`/`RIGHT` to reach the middle gem...

**Reply by Aaron-cly** [reply]
What? I thought the player stops at the boundary lol

**Reply by Derppening** [reply]
Right, I forgot about that (lol).

Originally I only expect students to check that the player can reach the gem without considering the inertia rules. If that's the case, I will revise the test cases to make it work if you only check for inertia rules as well :thinking:

**Reply by Aaron-cly** [reply]
Yeah thank you for the patience. I think it only makes sense when the game starts at a state where the player can actually win, that is all gems are reachable following the inertia rules. Thank you!

**Comment by Derppening** [comment, marked as answer]
This issue should be fixed in 7b1688d. 

Other implementations previously deemed as correct should continue to be correct.
